enterprise/server/util/oci/oci_test.go

```diff
@@ -1063,3 +1065,171 @@ func TestResolveImageDigest_AlreadyDigest_NoHTTPRequests(t *testing.T) {
+
+	resolver := newResolver(t, te)
+
+	{
+		counter.reset()
+		nameWithDigest, err := resolver.ResolveImageDigest(
+			context.Background(),
+			nameToResolve,
+			oci.RuntimePlatform(),
+			oci.Credentials{},
+		)
+		require.NoError(t, err)
+		resolvedDigest, err := name.NewDigest(nameWithDigest)
+		require.NoError(t, err)
+		require.Equal(t, pushedDigest.String(), resolvedDigest.DigestStr())
+
+		expectedRequests := map[string]int{
+			http.MethodGet + " /v2/":                                    1,
+			http.MethodHead + " /v2/" + imageName + "/manifests/latest": 1,
+		}
+		require.Empty(t, cmp.Diff(expectedRequests, counter.snapshot()))
+	}
+
+	{
+		counter.reset()
+		nameWithDigest, err := resolver.ResolveImageDigest(
+			context.Background(),
+			nameToResolve,
+			oci.RuntimePlatform(),
+			oci.Credentials{},
+		)
+		require.NoError(t, err)
+
+		resolvedDigest, err := name.NewDigest(nameWithDigest)
+		require.NoError(t, err)
+		require.Equal(t, pushedDigest.String(), resolvedDigest.DigestStr())
+
+		require.Empty(t, counter.snapshot())
+	}
+
+	{
+		ref, err := name.ParseReference(nameToResolve)
+		require.NoError(t, err)
+		registryAndRepoNoTag := ref.Context().String()
+		counter.reset()
+		nameWithDigest, err := resolver.ResolveImageDigest(
+			context.Background(),
+			registryAndRepoNoTag,
+			oci.RuntimePlatform(),
+			oci.Credentials{},
+		)
+		require.NoError(t, err)
+
+		resolvedDigest, err := name.NewDigest(nameWithDigest)
+		require.NoError(t, err)
+		require.Equal(t, pushedDigest.String(), resolvedDigest.DigestStr())
+
+		require.Empty(t, counter.snapshot())
+	}
+}
+
+func TestResolveImageDigest_CacheExpiration(t *testing.T) {
+	te := testenv.GetTestEnv(t)
+	flags.Set(t, "executor.container_registry_allowed_private_ips", []string{"127.0.0.1/32"})
+
+	counter := newRequestCounter()
+	registry := testregistry.Run(t, testregistry.Opts{
+		HttpInterceptor: func(w http.ResponseWriter, r *http.Request) bool {
+			counter.Inc(r)
+			return true
+		},
+	})
+
+	imageName := "cache_expiration"
+	_, img := registry.PushNamedImage(t, imageName)
+	pushedDigest, err := img.Digest()
+	require.NoError(t, err)
+	nameToResolve := registry.ImageAddress(imageName)
+
+	fakeClock := clockwork.NewFakeClock()
+	te.SetClock(fakeClock)
+	resolver := newResolver(t, te)
+
+	counter.reset()
+	nameWithDigest, err := resolver.ResolveImageDigest(
+		context.Background(),
+		nameToResolve,
+		oci.RuntimePlatform(),
+		oci.Credentials{},
+	)
+	require.NoError(t, err)
+	resolvedDigest, err := name.NewDigest(nameWithDigest)
+	require.NoError(t, err)
+	require.Equal(t, pushedDigest.String(), resolvedDigest.DigestStr())
+
+	expectedFirst := map[string]int{
+		http.MethodGet + " /v2/":                                    1,
+		http.MethodHead + " /v2/" + imageName + "/manifests/latest": 1,
+	}
+	require.Empty(t, cmp.Diff(expectedFirst, counter.snapshot()))
+
+	// 2) Immediate resolve should be a cache hit; no HTTP requests.
```

Comment:
nit: renumber comments starting from `1)`? or just delete the numbering to avoid hassle :P